buttonhub.py

```diff
@@ -161,54 +186,51 @@ def check_data_condition(condition, data, battery, device_id):
         else:
             current_part = current_part[key]
     if 'value' in condition:
-        passes = passes and check_condition(condition['value'], battery, device_id, current_part)
+        passes = passes and check_condition(condition['value'], context, carry_value=current_part)
     return passes
 
 
-def do_request(req, battery, device_id):
+def do_request(req, context):
     if 'delay' in req:
         time.sleep(req['delay'])
     if 'endpoints' in req:
         for sub_req in req['endpoints']:
-            do_request(sub_req, battery, device_id)
+            do_request(sub_req, context)
     if 'url' in req:
-        do_http(req, battery, device_id)
+        do_http(req, context)
     if 'topic' in req:
-        if client is None:
+        if mqtt_client is None:
             return
-        do_mqtt(req, battery, device_id, client)
+        do_mqtt(req, context, mqtt_client)
 
 
-def do_http(req, battery, device_id):
+def do_http(req, context):
     url = req.get('url')
     if not url:
         print("No URL defined")
-        return ""
-    if '{battery}' in url or '{device}' in url:
-        url = url.format(battery=str(battery), device=str(device_id))
+        return ''
+    url = url.format(**context)
```

Comment:
IIRC there was a reason why I conditionally applied the string.format() - it can break for some strings which contain {}, especially strings that contain json.

Reply:
You have to escape {}s if you use string.format() - I think my idea was that this way escaping is only needed when you actually *use* the templating feature